File: Array, List(python)/88. Merge Sorted Array .py
8.排序和搜索设计问题
本章包含了leetcode中的两节：排序和搜索，设计问题。因为这两节的题型比较相近，题目比较少，合并。
8.1合并两个有序数组

8.1合并两个有序数组
给定两个整数数组nums1和nums2，将nums2合并到nums1，使得nums1成为一个有序数组。

说明：初始化nums1和nums2的元素数量为m和n
·假设nums1有足够的空间（大于m+n），来保存nums2中的元素

输入 nums1=【1，2，3，0，0，0】，m=3
nums2=【2，5，6】 n=3
输出：【1，2，2，3，5，6】

8.1.2
解题思路：
1.pythonic算法
这一道题简单地说就是一道排序题，将两个列表合并成一个列表后排序，给一个整数列表排序，以nums1=【1，2，3，0，0，0】，m=3
nums2=【2，5，6】 n=3为例。

所要做的仅仅是合并nums1和nums2，然后使用python的列表sort函数而已。

2.详细算法
pythonic取巧，虽然达到了目的，但是并不清楚是怎么做到的。现在来看详细的算法过程，没错还是排序。将两个列表合并，和插入排序很像。

与正规的插入排序相比，有序数组的插入不是从第二个数开始插入的，可以直接从第二个有序数组的组头开始插入，因为数组是有序的。

8.1.3解题代码

1.pythonic
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        for i in range(n):
        	nums1[m+i]=nums2[i]
        nums1.sort()
        return

2.详细代码
合并有序数组的详细代码：
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        nums1[m:m+n]=nums2[:n]

        for right in range(m,m+n):
        	target=nums1[right]
        	for left in range(0, right):
        		if target<=nums1[left]:
        			nums1[left+1:right+1]=nums1[left:right] #使用python切片赋值
        			nums1[left]=target
        			break
        return

File: leetcode/Array, List(python)/88. Merge Sorted Array.java
//88You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n, representing the number of elements in nums1 and nums2 respectively.

Merge nums1 and nums2 into a single array sorted in non-decreasing order.

The final sorted array should not be returned by the function, but instead be stored inside the array nums1. To accommodate this, nums1 has a length of m + n, where the first m elements denote the elements that should be merged, and the last n elements are set to 0 and should be ignored. nums2 has a length of n.

 
/*
Example 1:

Input: nums1 = [1,2,3,0,0,0], m = 3, nums2 = [2,5,6], n = 3
Output: [1,2,2,3,5,6]
Explanation: The arrays we are merging are [1,2,3] and [2,5,6].
The result of the merge is [1,2,2,3,5,6] with the underlined elements coming from nums1.
Example 2:

Input: nums1 = [1], m = 1, nums2 = [], n = 0
Output: [1]
Explanation: The arrays we are merging are [1] and [].
The result of the merge is [1].
Example 3:

Input: nums1 = [0], m = 0, nums2 = [1], n = 1
Output: [1]
Explanation: The arrays we are merging are [] and [1].
The result of the merge is [1].
Note that because m = 0, there are no elements in nums1. The 0 is only there to ensure the merge result can fit in nums1.
 

Constraints:

nums1.length == m + n
nums2.length == n
0 <= m, n <= 200
1 <= m + n <= 200
-109 <= nums1[i], nums2[j] <= 109
 

Follow up: Can you come up with an algorithm that runs in O(m + n) time?
*/
最直观的方法是先将数组 nums2 放进数组 nums1 的尾部，然后直接对整个数组进行排序。

class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        for(int i=0;i!=n;++i){
        	nums1[m+i]=nums2[i];
        }
        Arrays.sort(nums1);//java内置数组排序是快速排序，时间复杂度空间复杂度都是nlog(n)
    }
}
复杂度分析

时间复杂度：O((m+n)\log(m+n))O((m+n)log(m+n))。
排序序列长度为 m+nm+n，套用快速排序的时间复杂度即可，平均情况为 O((m+n)\log(m+n))O((m+n)log(m+n))。

空间复杂度：O(\log(m+n))O(log(m+n))。
排序序列长度为 m+nm+n，套用快速排序的空间复杂度即可，平均情况为 O(\log(m+n))O(log(m+n))。

//方法2 双指针
//方法一没有利用数组 nums1 与 nums2已经被排序的性质。为了利用这一性质，我们可以使用双指针方法。这一方法将两个数组看作队列，每次从两个数组头部取出比较小的数字放到结果中。如下面的动画所示：
//我们为两个数组分别设置一个指针 p1 与 p2来作为队列的头部指针。代码实现如下：
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int p1 = 0, p2 = 0;
        int[] sorted = new int[m + n];
        int cur;
        while (p1 < m || p2 < n) {
            if (p1 == m) {
                cur = nums2[p2++];
            } else if (p2 == n) {
                cur = nums1[p1++];
            } else if (nums1[p1] < nums2[p2]) {
                cur = nums1[p1++];
            } else {
                cur = nums2[p2++];
            }
            sorted[p1 + p2 - 1] = cur;
        }
        for (int i = 0; i != m + n; ++i) {
            nums1[i] = sorted[i];
        }
    }
}


时间复杂度：O(m+n)O(m+n)。
指针移动单调递增，最多移动 m+nm+n 次，因此时间复杂度为 O(m+n)O(m+n)。

空间复杂度：O(m+n)O(m+n)。
需要建立长度为 m+nm+n 的中间数组 \textit{sorted}sorted。


//方法三：逆向双指针
//@source https://leetcode-cn.com/problems/merge-sorted-array/solution/he-bing-liang-ge-you-xu-shu-zu-by-leetco-rrb0/

//方法二中，之所以要使用临时变量，是因为如果直接合并到数nums1中，nums1中的元素可能会在取出之前被覆盖。那么如何直接避免覆盖 nums1中的元素呢？观察可知nums1的后半部分是空的，可以直接覆盖而不会影响结果。因此可以指针设置为从后向前遍历，每次取两者之中的较大者放进nums1的最后面。
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int p1 = m - 1, p2 = n - 1;
        int tail = m + n - 1;
        int cur;
        while (p1 >= 0 || p2 >= 0) {
            if (p1 == -1) {
                cur = nums2[p2--];
            } else if (p2 == -1) {
                cur = nums1[p1--];
            } else if (nums1[p1] > nums2[p2]) {
                cur = nums1[p1--];
            } else {
                cur = nums2[p2--];
            }
            nums1[tail--] = cur;
        }
    }
}

复杂度分析

时间复杂度：O(m+n)O(m+n)。
指针移动单调递减，最多移动 m+nm+n 次，因此时间复杂度为 O(m+n)O(m+n)。

空间复杂度：O(1)O(1)。
直接对数组 \textit{nums}_1nums 
1
​
原地修改，不需要额外空间。
